### ISSUE TITLE
Surefire fork failures with our provider

### ISSUE THREAD

**Issue by bartoszmajsak** [arquillian/smart-testing#124]
When the number of thread/surefire foks is set to more than one, and `reuseForks=true` then it fails with:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project arquillian-config-impl-base: ExecutionException There was an error in the forked process
[ERROR] java.util.NoSuchElementException: 
[ERROR] 	at org.apache.maven.surefire.booter.CommandReader$ClassesIterator.next(CommandReader.java:275)
[ERROR] 	at org.apache.maven.surefire.booter.CommandReader$ClassesIterator.next(CommandReader.java:248)
[ERROR] 	at org.apache.maven.surefire.booter.LazyTestsToRun$BlockingIterator.next(LazyTestsToRun.java:73)
[ERROR] 	at org.apache.maven.surefire.booter.LazyTestsToRun$BlockingIterator.next(LazyTestsToRun.java:61)
[ERROR] 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
[ERROR] 	at org.arquillian.smart.testing.surefire.provider.SmartTestingSurefireProvider.invoke(SmartTestingSurefireProvider.java:66)
[ERROR] 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
[ERROR] 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
[ERROR] 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
```

Does not matter which strategy is chosen.

**Comment by bartoszmajsak** [comment]
Reproducer test: https://github.com/arquillian/smart-testing/commit/69f39a1c090dd5a0a4025bbf81e7a534f3be2ef2#diff-3a2ffbf2f0baadc1295fcced8975ff13R46

**Comment by bartoszmajsak** [comment]
To further eliminate some vectors, I run the this problematic test against the repo dumped in the folder with

  * disabled `System.out` logging
  * using real Maven, not embedded one

This, unfortunately, still results in the same error, as you can see in this gist https://gist.github.com/bartoszmajsak/a2ad2e09b26120fa8a381c3ef957aec0

**Comment by bartoszmajsak** [comment]
Running against `maven-surefire-plugin:2.20` shows slightly different stack trace but apparently, provider is not happy being "prematurely" closed (meaning we pass less tests than originally calculated). 

https://gist.github.com/bartoszmajsak/b21e7bc281cc4da0ab626cfd46c83cfe

**Comment by bartoszmajsak** [comment]
Should be fixed by #127
